### PR TITLE
http signatures: add support for RSA message signing

### PR DESCRIPTION
**Pull request by ewilde**
TODO:

- [x] mac osx testing
- [x] linux testing
- [ ] arm

## Description
Updates deployment script to create keys for http-signatures

## Motivation and Context
Resolves issue #1103

## How Has This Been Tested?
- [x] **Mac (mojave)**

Output running on a system not running OpenFaaS:
```
$ ./deploy_stack.sh                                                                                                                          
Attempting to create credentials for gateway..
5kt9oxodp7xyz5td0lkxlzrin
yvrtnodgstfkukaeyga4vs4tr
[Credentials]
 username: admin 
 password: b676451f42836f2d39d7012d51714c64bde4630c6b022708285e39f776608c68
 echo -n b676451f42836f2d39d7012d51714c64bde4630c6b022708285e39f776608c68 | faas-cli login --username=admin --password-stdin

Enabling basic authentication for gateway..

Http encryption settings enabled

Deploying OpenFaaS core services
Creating service func_nats
Creating service func_queue-worker
Creating service func_prometheus
Creating service func_alertmanager
Creating service func_gateway
Creating service func_faas-swarm

```

**Requesting a known public key**

```
$ curl localhost:8080/certificates/callback -s | jq '.pem' -r                                                                   

-----BEGIN PUBLIC KEY-----
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA3tjZ3yP8DzZTKSVDuNKV
HZJWqhdVWlwuqFdOwJ+ArL65BLEH+aVO+jXjI2IoQDrGgg5dw1zdCiGundL/SMLa
4ABz2wAIbEEafvPMp7gxxfrPYb7Cy0mkEADZGSUdE7oOCJzkzUsbRFeqayyG8fm4
OOtqcE0Z/Yf9QF1lebOYTMQFhig9c40v40ewnxLucxkJEcYuNcg/SNyzRlbK7AKg
pGejhkWRU16FTRkR6C8caAj0MUkITqhSp7N4VOU/K5ZOwO6n6vyiza81lbNPSRG9
qCHLy+IppN0KC0SMhfQ5Kd+3U+LEHaCAJ5p7KnrtFAfAEfCcVRyKMXrO2RFXKUPW
1QIDAQAB
-----END PUBLIC KEY-----
```

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`

### CODE REVIEW

**Comment by alexellis** [comment]
This bash change looks like something that should just be done separately?

**Comment by alexellis** [comment]
Hi @ewilde 

Is `/certificates/callback` a standard documented somewhere?

Should we just use `/certificates/<SOMETHING_ELSE_MORE_GENERIC>`?

Alex

**Comment by ewilde** [comment]
> Hi @ewilde
> 
> Is `/certificates/callback` a standard documented somewhere?
> 
> Should we just use `/certificates/<SOMETHING_ELSE_MORE_GENERIC>`?
> 
> Alex

@alexellis no  `/certificates/callback` is not a standard. The idea was that you may host more certificates for signing other parts of the system. Hence `callback` for the webhooks.

**Comment by ewilde** [comment]
> This bash change looks like something that should just be done separately?

Remove bash changes

**Comment by ewilde** [comment]
@alexellis PTAL this is ready for a review

File: gateway/handlers/certificates_handler.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+type SecretsReader interface {
+	Read(key string) (string, error)
+}
+
+type FileSecretsReader struct {
+	SecretMountPath string
+}
+
+func (f *FileSecretsReader) Read(key string) (string, error) {
```
Comment (alexellis) [inline gateway/handlers/certificates_handler.go]: Reminds me of https://github.com/openfaas/openfaas-cloud/blob/master/sdk/secrets.go#L13

Comment (alexellis) [reply gateway/handlers/certificates_handler.go]: I think there are some other similar bits of code in the project for reading secrets, can you have a quick look?

File: gateway/handlers/certificates_handler_test.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+func TestMakeCertificatesHandler(t *testing.T) {
+	tests := []struct {
+		name   string
+		keyID  string
+		eval   func(r *http.Response) error
+		reader SecretsReader
+	}{
+		{
+			name:  "Get certificate",
```
Comment (alexellis) [inline gateway/handlers/certificates_handler_test.go]: `Get certificate that exists`?

File: gateway/handlers/certificates_handler_test.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+				}
+				return nil
+			},
+			reader: &TestSecretsReader{readCallBack: func(key string) (s string, e error) {
+				return testPublicKey, nil
+			}},
+		},
+		{
+
+			name:  "Unable to find certificate",
```
Comment (alexellis) [inline gateway/handlers/certificates_handler_test.go]: `Attempt to get certificate that doesn't exist`?

**Comment by burtonr** [comment]
Hey @ewilde! I know this has been sitting for a bit, and now there are some conflicts. Also, there were a few comments by Alex that have not yet been addressed. 

Could you give an update here please? I'd be willing to setup and do some testing on ARM if you don't have any devices available.

**Comment by alexellis** [comment]
@ewilde how is this affected now that the async call flows as follows:

GW (async path) -> NATS -> queue-worker -> GW (sync path)

Could you test when you have time?

**Comment by alexellis** [comment]
So that this work doesn't get missed, is it also on the 2019 roadmap? https://trello.com/b/qqjcNqs4/2019-openfaas-roadmap

File: gateway/handlers/certificates_handler_test.go (hunk 2 above)
Comment (ewilde) [reply gateway/handlers/certificates_handler_test.go]: okay

File: gateway/handlers/certificates_handler_test.go (hunk 3 above)
Comment (ewilde) [reply gateway/handlers/certificates_handler_test.go]: okay

File: gateway/handlers/certificates_handler.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+	value, err := ioutil.ReadFile(certificatePath)
+	if err != nil {
+		return "", fmt.Errorf("error reading find secret %s", certificatePath)
+	}
+
+	return string(value), nil
+}
+
+func MakeCertificatesHandler(reader SecretsReader) http.HandlerFunc {
```
Comment (alexellis) [inline gateway/handlers/certificates_handler.go]: Should this be served on `/.well-known/`?

**Comment by alexellis** [comment]
I still want to solve this problem, but am closing this PR due to its age.

**Pull request closed without merging**